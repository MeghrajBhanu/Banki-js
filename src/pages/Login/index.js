import React, { useState, useEffect } from "react";
import { useDispatch, useSelector } from "react-redux";
import { login } from "../../redux/actions/auth";
import { useNavigate } from "react-router";
import { clearMessage } from "../../redux/actions/message";
import {
  validateEmail,
  validatePan,
  validatePass,
} from "../../utils/InputValidations";
import { Outlet } from "react-router-dom";
/**
 * Login component handles user Login
 *
 * @returns {React.ReactElement}
 */
const Login = () => {
  useEffect(() => {
    dispatch(clearMessage());
  }, []);
  const [valuess, setValues] = useState({
    email: "",
    password: "",
    pancard: "",
  });
  const [successful, setSuccessful] = useState(false);
  const handleChange = (e) => {
    setValues({ ...valuess, [e.target.name]: e.target.value });
    handleInputValidation(e);
  };
  const { message } = useSelector((state) => state.message);
  const isLoggedIn = useSelector((state) => state.auth.isLoggedIn);
  const dispatch = useDispatch();
  const navigate = useNavigate();

  const [disable, setDisable] = useState(true);
  const [isEmailValid, setISEmailValid] = useState({
    isInputValid: false,
    errorMessage: "",
  });
  const [isPanValid, setISPanValid] = useState({
    isInputValid: false,
    errorMessage: "",
  });
  const [isPassValid, setISPassValid] = useState({
    isInputValid: false,
    errorMessage: "",
  });
  const handleInputValidation = (event) => {
    if (event.target.name === "email") {
      const { isInputValid, errorMessage } = validateEmail(event.target.value);
      setISEmailValid({
        isInputValid: isInputValid,
        errorMessage: errorMessage,
      });
    }
    if (event.target.name === "pancard") {
      const { isInputValid, errorMessage } = validatePan(event.target.value);
      setISPanValid({
        isInputValid: isInputValid,
        errorMessage: errorMessage,
      });
    }
    if (event.target.name === "password") {
      const { isInputValid, errorMessage } = validatePass(event.target.value);
      setISPassValid({
        isInputValid: isInputValid,
        errorMessage: errorMessage,
      });
    }
    if (
      !isEmailValid.isInputValid ||
      !isPanValid.isInputValid ||
      !isPassValid.isInputValid
    ) {
      setDisable(true);
    } else {
      setDisable(false);
    }
  };
  const onSubmit = (e) => {
    e.preventDefault();

    dispatch(login(valuess))
      .then((data) => {
        setSuccessful(true);

        navigate("/login/landing");
      })
      .catch(() => {
        setSuccessful(false);
      });
    setDisable(true);
  };
  return (
    <>
      {isLoggedIn && <Outlet />}
      {!isLoggedIn && (
        <section className="vh-100" style={{ backgroundColor: "#eee" }}>
          <div className="container h-100">
            <div className="row d-flex justify-content-center align-items-center h-100">
              <div className="col-lg-12 col-xl-11">
                <div
                  className="card text-black"
                  style={{ borderRadius: "25px" }}
                >
                  <div className="card-body p-md-5">
                    <div className="row justify-content-center">
                      <div className="col-md-10 col-lg-6 col-xl-5 order-2 order-lg-1">
                        <p className="text-center h1 fw-bold mb-5 mx-1 mx-md-4 mt-4">
                          Login
                        </p>

                        <form className="mx-1 mx-md-4 ">
                          <div className="d-flex flex-row align-items-center mb-3">
                            <i className="fas fa-envelope fa-lg me-3 fa-fw"></i>
                            <div className="form-outline flex-fill mb-0">
                              <input
                                type="email"
                                aria-label="email"
                                name="email"
                                value={valuess.email}
                                id="form3Example3c"
                                className="form-control"
                                placeholder="Your Email"
                                onChange={handleChange}
                                onBlur={handleInputValidation}
                                required="required"
                              />
                            </div>
                          </div>
                          {!isEmailValid.isInputValid ? (
                            <div
                              className="text-danger mb-3"
                              style={{ marginLeft: "50px" }}
                            >
                              {isEmailValid.errorMessage}
                            </div>
                          ) : null}
                          <div className="d-flex flex-row align-items-center mb-4">
                            <i className="fas fa-solid fa-id-card fa-lg me-3 fa-fw "></i>
                            <div className="form-outline flex-fill mb-0">
                              <input
                                type="text"
                                aria-label="pancard"
                                name="pancard"
                                value={valuess.pancard}
                                id="form3Example4cd"
                                className="form-control"
                                placeholder="Pancard"
                                onChange={handleChange}
                                onBlur={handleInputValidation}
                                required="required"
                              />
                            </div>
                          </div>
                          {!isPanValid.isInputValid ? (
                            <div
                              className="text-danger mb-3"
                              style={{ marginLeft: "50px" }}
                            >
                              {isPanValid.errorMessage}
                            </div>
                          ) : null}

                          <div className="d-flex flex-row align-items-center mb-4">
                            <i className="fas fa-key fa-lg me-3 fa-fw"></i>

                            <div className="form-outline flex-fill mb-0">
                              <input
                                type="password"
                                aria-label="password"
                                name="password"
                                id="form3Example4c"
                                className="form-control"
                                placeholder="Password"
                                value={valuess.password}
                                onChange={handleChange}
                                onBlur={handleInputValidation}
                                required="required"
                              />
                            </div>
                          </div>
                          {/* <small
                            id="passwordHelpBlock"
                            class="form-text text-muted"
                          >
                            Your password must be 8-20 characters long, contain
                            letters and numbers, and must not contain spaces,
                            special characters, or emoji.
                          </small> */}
                          {!isPassValid.isInputValid ? (
                            <div
                              className="text-danger mb-3"
                              style={{ marginLeft: "50px" }}
                            >
                              {isPassValid.errorMessage}
                            </div>
                          ) : null}

                          <div className="d-flex justify-content-center mx-4 mb-3 mb-lg-4">
                            <button
                              type="button"
                              className="btn btn-primary btn-lg"
                              onClick={onSubmit}
                              disabled={disable}
                            >
                              Login
                            </button>
                          </div>
                        </form>
                        {message && (
                          <div className="form-group">
                            <div
                              className={
                                successful
                                  ? "alert alert-success"
                                  : "alert alert-danger"
                              }
                              role="alert"
                            >
                              {message}
                            </div>
                          </div>
                        )}
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </section>
      )}
    </>
  );
};
export default Login;
