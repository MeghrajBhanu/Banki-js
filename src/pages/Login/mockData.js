export const mockLoginData = {
    name: 'mock user',
    email: 'mockuser@gmail.com',
    password: 'Qwety@123',
    pancard: 'RQTYI1234P',
  };
  export const mock_Invalid_LoginData = {
    name: 'm',
    email: 'mOCK1',
    password: 'ui',
    pancard: '1234',
  };