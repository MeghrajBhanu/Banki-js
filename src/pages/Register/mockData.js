export const mockRegisterData = {
    name: 'mock user',
    email: 'mockuser@gmail.com',
    password: 'Qwety@123',
    confirmpassword: 'Qwety@123',
    pancard: 'RQTYI1234P',
  };
  export const mock_Invalid_RegisterData = {
    name: 'm',
    email: 'mOCK1',
    password: 'ui',
    confirmpassword: 'Qwety@123',
    pancard: '1234',
  };