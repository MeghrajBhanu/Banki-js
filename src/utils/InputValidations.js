export const validateEmail = (checkingText) => {
    
    const regexp = /^\w+([\.-]?\w+)*@\w+([\.-]?\w+)*(\.\w{2,3})+$/; 
   
    
    if (regexp.exec(checkingText) !== null) {
            return {
                isInputValid: true,
                errorMessage: ''
            };
        } else {
            return {
                isInputValid: false,
                errorMessage: 'Invalid Email Address'
            };
        }
}
export const validatePan = (checkingText) => {
    
    
    if (checkingText.length === 10) {
            return {
                isInputValid: true,
                errorMessage: ''
            };
        } else {
            return {
                isInputValid: false,
                errorMessage: 'Invalid Pan Number'
            };
        }
}
export const validatePass = (checkingText) => {
    
    
    if (checkingText!=="") {
            return {
                isInputValid: true,
                errorMessage: ''
            };
        } else {
            return {
                isInputValid: false,
                errorMessage: 'Invalid Password'
            };
        }
}