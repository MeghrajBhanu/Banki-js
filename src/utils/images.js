import bank1 from "../assets/bank1.jpg";
import bank2 from "../assets/bank2.png";
import bank3 from "../assets/bank3.jfif";
import bank4 from "../assets/bank4.png";

const image_map={
    "Bank1":bank1,
    "Bank2":bank2,
    "Bank3":bank3,
    "Bank4":bank4,
}
export default image_map;