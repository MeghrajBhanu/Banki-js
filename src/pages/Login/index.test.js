import { render, screen } from "@testing-library/react";
import { renderComponent,onlywrap } from "../../utils/test_helper";
import Login from ".";
import userEvent from "@testing-library/user-event";
import { mockLoginData, mock_Invalid_LoginData } from "./mockData";
import renderer from "react-test-renderer";
const user = userEvent.setup();
describe("Login Page", () => {
  test("should show Login heading", () => {
    renderComponent(<Login />);
    const value = screen.getAllByText(/Login/i);
    expect(value[0]).toBeInTheDocument();
  });

  test("should render the initial form", () => {
    renderComponent(<Login />);
    expect(screen.getByRole("textbox", { name: "email" })).toBeInTheDocument();
    expect(
      screen.getByRole("textbox", { name: "pancard" })
    ).toBeInTheDocument();
    expect(screen.getByPlaceholderText("Password")).toBeInTheDocument();
    expect(screen.getByRole("button", { name: "Login" })).toBeInTheDocument();
  });
  test("should handle input changes", async () => {
    renderComponent(<Login />);
    const email = screen.getByRole("textbox", { name: "email" });
    const password = screen.getByPlaceholderText("Password");
    const pancard = screen.getByRole("textbox", { name: "pancard" });

    await user.type(email, mockLoginData.email);
    expect(email).toHaveValue(mockLoginData.email);

    await user.type(password, mockLoginData.password);
    expect(password).toHaveAttribute("type", "password");
    expect(password).toHaveValue(mockLoginData.password);

    await user.type(pancard, mockLoginData.pancard);
    expect(pancard).toHaveValue(mockLoginData.pancard);
  });
  test("should handle errors in input", async () => {
    renderComponent(<Login />);
    
    const email = screen.getByRole("textbox", { name: "email" });
    const password = screen.getByPlaceholderText("Password");
    const pancard = screen.getByRole("textbox", { name: "pancard" });

    email.focus();
    await user.tab();
    expect(screen.getByText("Email Address is required")).toBeInTheDocument();

    password.focus();
    await user.tab();
    expect(screen.getByText("Password Required")).toBeInTheDocument();

    pancard.focus();
    await user.tab();
    expect(screen.getByText("Pan Number is required")).toBeInTheDocument();

    await user.type(email, mock_Invalid_LoginData.email);
    expect(screen.getByText("Invalid Email Address")).toBeInTheDocument();

    await user.type(password, mock_Invalid_LoginData.password);
    expect(screen.getByText("Invalid Password")).toBeInTheDocument();

    await user.type(pancard, mock_Invalid_LoginData.pancard);
    expect(screen.getByText("Invalid Pan Number")).toBeInTheDocument();
  });
  test("should handle unsucessfull login(email)",async()=>{
    let {store}=renderComponent(<Login/>);
    const email = screen.getByRole("textbox", { name: "email" });
    const password = screen.getByPlaceholderText("Password");
    const pancard = screen.getByRole("textbox", { name: "pancard" });
    const login=screen.getByRole('button', { name: /login/i });
     
    await user.type(email, mockLoginData.email);
    await user.type(password, mockLoginData.password);
    await user.type(pancard, "POIYU1234L");
    await user.click(document.body);
    expect(login).toBeEnabled()
    await user.click(login);
    
    expect(store.getState().auth.token).toBe(null);  
  })
  test("should handle sucessfull Login",async()=>{
    
    let {store}=renderComponent(<Login />);
    screen.logTestingPlaygroundURL()
    const email = screen.getByRole("textbox", { name: "email" });
    const password = screen.getByPlaceholderText("Password");
    const pancard = screen.getByRole("textbox", { name: "pancard" });
    const login=screen.getByRole('button', { name: /login/i });
    
    
    await user.type(email, mockLoginData.email);
    await user.type(password, mockLoginData.password);
    await user.type(pancard, mockLoginData.pancard);
    await user.click(document.body);
    expect(login).toBeEnabled()
    await user.click(login);
    
    expect(store.getState().auth.token).not.toBe(null);
    
    

  })
  test("snapshot test", () => {
    const tree = renderer.create(onlywrap(<Login />)).toJSON();
    expect(tree).toMatchSnapshot();
  });
  
});
