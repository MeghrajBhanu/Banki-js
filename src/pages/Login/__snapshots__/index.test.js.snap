// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login Page snapshot test 1`] = `null`;
